crypto/cipher_extra/e_chacha20poly1305.c

```diff
@@ -346,3 +382,317 @@ const EVP_AEAD *EVP_aead_chacha20_poly1305(void) {
+    }
+  }
+  if (counter_nonce) {
+    for (i = 0; i < CHACHA_CTR_IV_LEN / 4; i++) {
+      key->counter_nonce[i] = *(((const uint32_t *) counter_nonce) + i);
+    }
+  }
+  key->partial_len = 0;
+  return 1;
+}
+
+static int cipher_chacha20_poly1305_init(EVP_CIPHER_CTX *ctx,
+                                         const uint8_t *key,
+                                         const uint8_t *iv, int32_t enc) {
+  CIPHER_CHACHA_POLY_CTX *cipher_ctx = CCP_CTX(ctx);
+  if (!key && !iv) {
+    return 1;
+  }
+  cipher_ctx->len.aad = 0;
+  cipher_ctx->len.text = 0;
+  cipher_ctx->pad_aad = 0;
+  cipher_ctx->poly_initialized = 0;
+  // Init can be called multiple times before starting the cipher to
+  // independently initialize any combination of Key/IV/NULL.
+  if (iv != NULL) {
+    // Start the counter at 0 and copy over the nonce(iv)
+    uint8_t counter_nonce[CHACHA_CTR_IV_LEN] = {0};
+    memcpy(counter_nonce + CHACHA_CTR_IV_LEN - CHACHA_IV_LEN, iv,
+           CHACHA_IV_LEN);
+    cipher_chacha20_poly1305_init_key(cipher_ctx, key, counter_nonce);
+    // Nonce occupies the last 3 indices of the array
+    cipher_ctx->iv[0] = cipher_ctx->key.counter_nonce[1];
+    cipher_ctx->iv[1] = cipher_ctx->key.counter_nonce[2];
+    cipher_ctx->iv[2] = cipher_ctx->key.counter_nonce[3];
+  } else {
+    cipher_chacha20_poly1305_init_key(cipher_ctx, key, NULL);
+  }
+  return 1;
+}
+
+static int cipher_chacha20_do_cipher(EVP_CIPHER_CTX *ctx, uint8_t *out,
+                                     const uint8_t *inp, size_t len)
+{
+  CIPHER_CHACHA_POLY_CTX  *cipher_ctx = CCP_CTX(ctx);
+  CIPHER_CHACHA_KEY *key = CC_KEY(cipher_ctx);
+  uint32_t n, rem, ctr32;
+
+  // Complete any partial block
+  if ((n = key->partial_len)) {
+    // Compute the cipher using our partially used key and any new input up
+    // to the next block
+    while (len && n < CHACHA_BLOCK_LEN) {
+      *out++ = *inp++ ^ key->buf[n++];
+      len--;
+    }
+    key->partial_len = n;
+
+    // If we consumed all the input, we're done
+    if (len == 0) {
+      return 1;
+    }
+
+    // If we completed a block, increment the counter
+    if (n == CHACHA_BLOCK_LEN) {
+      key->partial_len = 0;
+      key->counter_nonce[0]++;
+    }
+  }
+
+  // Truncate down to the last complete block prior to the bulk cipher
+  rem = (uint32_t)(len % CHACHA_BLOCK_LEN);
+  len -= rem;
+  ctr32 = key->counter_nonce[0];
+  while (len >= CHACHA_BLOCK_LEN) {
+    size_t blocks = len / CHACHA_BLOCK_LEN;
+    // 1<<28 is just a not-so-small yet not-so-large number... Below
+    // condition is practically never met, but it has to be checked for code
+    // correctness.
+    if (sizeof(size_t) > sizeof(uint32_t) && blocks > (1U<<28)) {
+      blocks = (1U << 28);
+    }
+
+    // As ChaCha20_ctr32 operates on 32-bit counter, caller has to handle
+    // overflow. 'if' below detects the overflow, which is then handled by
+    // limiting the amount of blocks to the exact overflow point...
```

Comment:
What does the caller need to do? Does this need to change the documentation for any of the EVP_update/encrypt functions? Only got to this point.

Reply:
The caller has to update their code to not do >256GB of data in a single ChaCha20-Poly1305 cipher. I can drop a note in the header file.